### ISSUE TITLE
[🐛] getInitialNotification returns undefined

### ISSUE THREAD

**Issue by bsonmez** [invertase/react-native-firebase#4446]
## Issue

I used to have installed those packages below. Then I had to install Crashlytics, so I had to upgrade those with the latest version.

After, I started not getting getInitialNotification on the foreground, returned undefined initial, and not triggered after once.

Only changes are made by installing/upgrading packages; there are not any other changes in the codebase. When I downgraded its old versions,  it started working as it should in iOS but not in Android.

I deleted node_modules, Pods and `./gradlew clean` in every new setup.

**Before** (It was working perfectly)
```json
"@react-native-firebase/analytics": "7.0.1",
"@react-native-firebase/app": "7.0.1",
"@react-native-firebase/messaging": "7.0.1",
```

**With Crashlytics** (getInitialNotification does not work in this group)

```json
"@react-native-firebase/analytics": "7.6.8",
"@react-native-firebase/app": "8.4.6",
"@react-native-firebase/crashlytics": "^8.4.11",
"@react-native-firebase/messaging": "7.9.1",
```

**After** (Partially Works)
In iOS getInitialNotification works. Also receives notifications in background and when its killed.
In Android foreground and killed doesn't work but background works,

```json
"@react-native-firebase/analytics": "7.1.7",
"@react-native-firebase/app": "8.0.0",
"@react-native-firebase/crashlytics": "8.0.0",
"@react-native-firebase/messaging": "7.2.0",
```

**TLDR;** 

getInitialNotification returns undefined in both Android and iOS in the versions above and does not trigger after once at all.

---

### Javascript

<details><summary>Click To Expand</summary>
<p>

#### `package.json`:

```json
"@react-native-firebase/analytics": "7.6.8",
"@react-native-firebase/app": "8.4.6",
"@react-native-firebase/crashlytics": "^8.4.11",
"@react-native-firebase/messaging": "7.9.1",
```

#### `firebase.json` for react-native-firebase v6:

```json
`{
  "react-native": {
    "crashlytics_debug_enabled": true,
    "crashlytics_auto_collection_enabled": true,
    "crashlytics_disable_auto_disabler": true
  }
}
`
```

</details>

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
- Firebase/Analytics (6.25.0):
    - Firebase/Core
  - Firebase/Core (6.25.0):
    - Firebase/CoreOnly
    - FirebaseAnalytics (= 6.5.1)
  - Firebase/CoreOnly (6.25.0):
    - FirebaseCore (= 6.7.1)
  - Firebase/Crashlytics (6.25.0):
    - Firebase/CoreOnly
    - FirebaseCrashlytics (~> 4.1.0)
  - Firebase/Messaging (6.25.0):
    - Firebase/CoreOnly
    - FirebaseMessaging (~> 4.4.1)
  - FirebaseAnalytics (6.5.1):
    - FirebaseCore (~> 6.7)
    - FirebaseInstallations (~> 1.2)
    - GoogleAppMeasurement (= 6.5.1)
    - GoogleUtilities/AppDelegateSwizzler (~> 6.0)
    - GoogleUtilities/MethodSwizzler (~> 6.0)
    - GoogleUtilities/Network (~> 6.0)
    - "GoogleUtilities/NSData+zlib (~> 6.0)"
    - nanopb (~> 1.30905.0)
  - FirebaseAnalyticsInterop (1.5.0)
  - FirebaseCore (6.7.1):
    - FirebaseCoreDiagnostics (~> 1.3)
    - FirebaseCoreDiagnosticsInterop (~> 1.2)
    - GoogleUtilities/Environment (~> 6.5)
    - GoogleUtilities/Logger (~> 6.5)
  - FirebaseCoreDiagnostics (1.4.0):
    - GoogleDataTransportCCTSupport (~> 3.1)
    - GoogleUtilities/Environment (~> 6.5)
    - GoogleUtilities/Logger (~> 6.5)
    - nanopb (~> 1.30905.0)
  - FirebaseCoreDiagnosticsInterop (1.2.0)
  - FirebaseCrashlytics (4.1.1):
    - FirebaseAnalyticsInterop (~> 1.2)
    - FirebaseCore (~> 6.6)
    - FirebaseInstallations (~> 1.1)
    - GoogleDataTransport (~> 6.1)
    - GoogleDataTransportCCTSupport (~> 3.1)
    - nanopb (~> 1.30905.0)
    - PromisesObjC (~> 1.2)
  - FirebaseInstallations (1.3.0):
    - FirebaseCore (~> 6.6)
    - GoogleUtilities/Environment (~> 6.6)
    - GoogleUtilities/UserDefaults (~> 6.6)
    - PromisesObjC (~> 1.2)
  - FirebaseInstanceID (4.3.4):
    - FirebaseCore (~> 6.6)
    - FirebaseInstallations (~> 1.0)
    - GoogleUtilities/Environment (~> 6.5)
    - GoogleUtilities/UserDefaults (~> 6.5)
  - FirebaseMessaging (4.4.1):
    - FirebaseAnalyticsInterop (~> 1.5)
    - FirebaseCore (~> 6.6)
    - FirebaseInstanceID (~> 4.3)
    - GoogleUtilities/AppDelegateSwizzler (~> 6.5)
    - GoogleUtilities/Environment (~> 6.5)
    - GoogleUtilities/Reachability (~> 6.5)
    - GoogleUtilities/UserDefaults (~> 6.5)
- GoogleAppMeasurement (6.5.1):
    - GoogleUtilities/AppDelegateSwizzler (~> 6.0)
    - GoogleUtilities/MethodSwizzler (~> 6.0)
    - GoogleUtilities/Network (~> 6.0)
    - "GoogleUtilities/NSData+zlib (~> 6.0)"
    - nanopb (~> 1.30905.0)
  - GoogleDataTransport (6.2.1)
  - GoogleDataTransportCCTSupport (3.2.0):
    - GoogleDataTransport (~> 6.1)
    - nanopb (~> 1.30905.0)
  - GoogleUtilities/AppDelegateSwizzler (6.7.2):
    - GoogleUtilities/Environment
    - GoogleUtilities/Logger
    - GoogleUtilities/Network
  - GoogleUtilities/Environment (6.7.2):
    - PromisesObjC (~> 1.2)
  - GoogleUtilities/Logger (6.7.2):
    - GoogleUtilities/Environment
  - GoogleUtilities/MethodSwizzler (6.7.2):
    - GoogleUtilities/Logger
  - GoogleUtilities/Network (6.7.2):
    - GoogleUtilities/Logger
    - "GoogleUtilities/NSData+zlib"
    - GoogleUtilities/Reachability
  - "GoogleUtilities/NSData+zlib (6.7.2)"
  - GoogleUtilities/Reachability (6.7.2):
    - GoogleUtilities/Logger
  - GoogleUtilities/UserDefaults (6.7.2):
    - GoogleUtilities/Logger
  - iCarousel (1.8.3)
  - InsiderHybrid (1.1.0)
  - InsiderMobile (10.4.0)
  - InsiderMobileAdvancedNotification (1.0.5):
    - iCarousel
  - nanopb (1.30905.0):
    - nanopb/decode (= 1.30905.0)
    - nanopb/encode (= 1.30905.0)
  - nanopb/decode (1.30905.0)
  - nanopb/encode (1.30905.0)
  - RNFBAnalytics (7.1.7):
    - Firebase/Analytics (~> 6.25.0)
    - React
    - RNFBApp
  - RNFBApp (8.0.0):
    - Firebase/CoreOnly (~> 6.25.0)
    - React
  - RNFBCrashlytics (8.0.0):
    - Firebase/Core (~> 6.25.0)
    - Firebase/Crashlytics (~> 6.25.0)
    - React
    - RNFBApp
  - RNFBMessaging (7.2.0):
    - Firebase/Analytics (~> 6.25.0)
    - Firebase/Messaging (~> 6.25.0)
    - React
    - RNFBApp
```

#### `AppDelegate.m`:

```objc
// N/A
```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [ ] my application is an AndroidX application?
- [ ] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [ ] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
buildscript {
    ext {
        buildToolsVersion = "28.0.3"
        minSdkVersion = 16
        compileSdkVersion = 29
        targetSdkVersion = 29
    }
    repositories {
        google()
        jcenter()
        mavenCentral()
    }
    dependencies {
        classpath('com.android.tools.build:gradle:3.6.3')
        classpath('com.google.gms:google-services:4.3.3')
        classpath('com.google.firebase:firebase-crashlytics-gradle:2.2.0')
        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }
}

allprojects {
    repositories {
        google()
        jcenter()
        mavenLocal()
        maven {
            // All of React Native (JS, Obj-C sources, Android binaries) is installed from npm
            url("$rootDir/../node_modules/react-native/android")
        }
        maven {
            // Android JSC is installed from npm
            url("$rootDir/../node_modules/jsc-android/dist")
        }


        maven { url 'https://jitpack.io' }

    }
}

```

#### `android/app/build.gradle`:

```groovy
dependencies {
    implementation fileTree(dir: "libs", include: ["*.jar"])
    implementation "com.facebook.react:react-native:+"  // From node_modules
    implementation 'com.android.support:multidex:1.0.3'
    implementation 'androidx.appcompat:appcompat:1.2.0-beta01'
    implementation 'androidx.swiperefreshlayout:swiperefreshlayout:1.1.0-rc01'
    implementation 'com.facebook.android:facebook-android-sdk:5.15.3'
    implementation 'com.google.firebase:firebase-analytics:17.4.2'
    implementation 'com.google.firebase:firebase-messaging:20.2.0'
    implementation 'androidx.lifecycle:lifecycle-extensions:2.2.0'
    implementation 'com.google.android.gms:play-services-ads:19.1.0'

    if (enableHermes) {
        def hermesPath = "../../node_modules/hermes-engine/android/";
        debugImplementation files(hermesPath + "hermes-debug.aar")
        releaseImplementation files(hermesPath + "hermes-release.aar")
    } else {
        implementation jscFlavor
    }
}

apply plugin: 'com.google.gms.google-services'
apply plugin: 'com.google.firebase.crashlytics'
// Run this once to be able to run the application with BUCK
// puts all compile dependencies into folder libs for BUCK to use
task copyDownloadableDepsToLibs(type: Copy) {
    from configurations.compile
    into 'libs'
}
```
</p>
</details>

---

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

**Comment by mikehardy** [comment]
The API docs for v6+ I believe tell you that you must call getInitialNotification after a component mount (late in the startup sequence...) or it can return undefined. Can you post a stripped down App.js that reproduces the problem if dropped in to a sample project?

**Comment by bsonmez** [comment]
src/index.js
```js
  useEffect(() => {
    const unsubscribe = firebase
      .messaging()
      .onMessage(async (remoteMessage) => {
        if (remoteMessage) {
              console.log(remoteMessage)
        }
      });

    if (ref.current) {
      firebase
        .messaging()
        .getInitialNotification()
        .then((remoteMessage) => {
          if (remoteMessage) {
              console.log(remoteMessage)
          }
        });
    }

    return () => {
      unsubscribe();
    };
  }, [ref.current && ref.current.navigate ? "ready" : "not-ready", appState]);
```

/index.js

```js
messaging().setBackgroundMessageHandler(async (remoteMessage) => {
  console.log("Message handled in the background!", remoteMessage);
});
````

This piece of code was working perfectly on `7.0.1`. I believe the version is also v6 in `7.0.1`, so there should be no changes after updating packages unless the API didn't change.

**Comment by mikehardy** [comment]
I don't see any App rendered or component mounted, when does useEffect fire? an App.js would I presume render something

**Comment by bsonmez** [comment]
Sorry for the late response. In App.js we render providers for auth, tracking, and language providers, etc. Then we render `src/index.js`. Lıke I said, it was working like this. Only changes updating versions

App.js
```js
return (
    <StoreProvider>
        <ConfigProvider >
          <WorldProvider>
            <UIProvider>
              <ApolloProvider>
                <NotificationProvider>
                  <AuthProvider>
                    <TrackerProvider>
                      <ActionSheetProvider>
                        <BottomSheetProvider>
                           <Root /> <========= Here is src/index.js
                        </BottomSheetProvider>
                      </ActionSheetProvider>
                    </TrackerProvider>
                  </AuthProvider>
                </NotificationProvider>
              </ApolloProvider>
            </UIProvider>
          </WorldProvider>
        </ConfigProvider>
      </StoreProvider>
)
```

/index.js
```js
import { AppRegistry, YellowBox } from "react-native";
import React, { useEffect } from "react";
import { enableScreens } from "react-native-screens";
import App from "./App";
import { log } from "./src/constants";
import { name as appName } from "./app.json";
import messaging from "@react-native-firebase/messaging";

enableScreens();

messaging().setBackgroundMessageHandler(async (remoteMessage) => {
  console.log("Message handled in the background!", remoteMessage);
});

function HeadlessCheck({ isHeadless }) {
  useEffect(
    React.useCallback(() => {
      (async () => {
        const registration = await messaging().registerDeviceForRemoteMessages();
        log.info("MESSAGING REGISTERED", registration);
      })();
    }),
  );

  if (isHeadless) {
    // App has been launched in the background by iOS, ignore
    return null;
  }
  return <App />;
}

AppRegistry.registerComponent(appName, () => HeadlessCheck);
```

**Comment by mikehardy** [comment]
Sorry, you must realize that code is not executable. I'm talking about a *standalone* *runs with nothing other than the App.js posted* code we can drop in to reproduce. We don't have access to your code base.
